### ISSUE TITLE
[ALM Accelerator - QUESTION] New feature flow documentation

### ISSUE THREAD

**Issue by sefa-phaag** [microsoft/coe-starter-kit#4902]
### What is your question?

Is there any overall documentation for the branching flow the ALM Accelerator Canvas App is designed to follow? It's not clear to me what the "correct" way is to create the feature branch so that I end up with a source control branch and synchronized development environment (assuming it's supposed to work that way). It's also not clear to me how to manage a pull request to pull those changes back into the main solution (although part of that might be that I'm not super clear on the pull request pattern to begin with).

### What component are you experiencing the issue with?

ALM Accelerator Canvas App

### What solution version are you using?

February 2022

**Comment by mikefactorial** [comment]
@sefa-phaag pardon the delayed response. We're currently working on some documentation to make this more discoverable. Below is a rough draft of that documentation. Would love to get your feedback on this if you have any.

# ALM Accelerator for Power Platform branching and environment strategy

The ALM Accelerator provides a default environment and branching strategy that is designed to help you get started quickly. The default strategy is based on the [CoE Starter Kit](https://docs.microsoft.com/power-platform/guidance/coe/setup) development team’s ALM process. The default strategy is designed to be [customized to meet your organization’s needs](almaccelerator-byot.md), but customers who adopt the default strategy can use the out-of-the-box pipelines without any customization.

## Default branching strategy

### Initial Commit Branching

The first time a solution is committed using the ALM Accelerator app a maker will choose to create a new branch for their feature from the main branch. During the first commit 2 new branches are created.

![FirstRun](https://user-images.githubusercontent.com/42348035/225115802-e7ac9f47-b274-47cf-b0a8-37a028606961.png)

1. The maker's feature branch which is based on the main branch. The feature branch is where the maker will make changes to the solution until it is ready to be merged into the solution branch.
1. The solution branch which is based on the main branch. The solution branch will initially only contain the solution's deployment pipeline templates. The solution branch is where the maker will merge their feature branch when it is ready to be deployed to the validation environment.

### Feature Commit Branching

When subsequent features are created by the maker a **new feature branch** should be **created from the solution branch**.

![FeatureRun](https://user-images.githubusercontent.com/42348035/225115850-3ee1c844-fea1-4605-8a36-7e8e5283e52e.png)

## Default environment strategy

By default when you create a deployment profile using the ALM Accelerator app it will assume that you are using the following environment strategy and the branching strategy described above:

| Environment | Purpose | Notes |
| --- | --- | --- |
| Dev | Development environment for makers to test and validate changes | Makers can either work in a shared environment or their own isolated environment|
| Validation | Environment for validating that a solution built from source control can be deployed | Deployment to the validation environment happens automatically, based on a branch policy, when a new pull request is created from a maker's feature branch to the solution branch |
| Test | Environment for testing a solution built from source control | Deployment to the test environment happens automatically when a pull request is merged from a maker's feature branch to the solution branch |
| Production | Production environment for the solution | Deployment to the production environment happens automatically when a pull request is merged from the solution branch to the main branch |

**Comment by sefa-phaag** [comment]
Thanks, Mike. That's definitely a helpful overview. A couple of follow-up questions:

- What is your recommended repo structure? All solutions in a single repo as subfolders?
- If Validation, Test, and Production deployments happen automatically, what is the purpose of the Deploy functionality within the app and does that conflict? What are you supposed to chose as the source and target branches in the Advanced Maker experience? With just the Maker experience, you only choose a deployment profile. Are these flows supposed to be different?
- What is the "proper" process to set up a new development environment from within the ALM Accelerator app? If I created a brand new environment in Power Apps and used the native functionality to import a solution, it would prompt me to set up connections, set environment variables, etc., but importing from source control in the app doesn't do that. You can select a deployment profile when importing, but understandably don't seem to work when importing into an environment other than the one they were created for (since the selected connections don't actually exist).
- Do you recommend "feature" environments to correlate with [feature branches](https://learn.microsoft.com/en-us/azure/devops/repos/git/git-branching-guidance?view=azure-devops) or is it the intention that a maker would have a single environment and delete / re-import solutions as needed to essentially "check out" a feature branch if they are working on multiple simultaneously?
- On step 5, you mean once the pull request is merged into the main branch, not when the pull request is created, correct?
- Are these branch policies actually configured or are there additional steps that need to be completed to make the automatic deployment happen? I don't see any reference to this in the [ALM Accelerator pipeline documentation](https://learn.microsoft.com/en-us/power-platform/guidance/coe/almaccelerator-pipelines).

**Comment by mikefactorial** [comment]
@sefa-phaag thanks for the follow up questions

What is your recommended repo structure? All solutions in a single repo as subfolders?
 - We use a mono repo with subfolders internally for the CoE Kit, but the ALM Accelerator is flexible enough to allow for any scenario you choose based on your requirements. Since we're a single team working on the same project across many solutions this makes sense for us and is simpler to maintain without having to create a new repo manually for new solutions. However, each of our Kits (Creator, CoE, Automation) have separate projects and repos.

If Validation, Test, and Production deployments happen automatically, what is the purpose of the Deploy functionality within the app and does that conflict? 
 - The deploy button in the app is responsible for creating Pull Requests the deployments happen once the Pull Requests are approved in both the maker and the advanced maker scenario.
  
What are you supposed to choose as the source and target branches in the Advanced Maker experience? 
 - Source and Target branches are as described above (e.g. In step 3 the user is choosing their feature branch as source and solution branch as target)

With just the Maker experience, you only choose a deployment profile. Are these flows supposed to be different?
 - The flows are the same they both create Pull Requests based on source and target branch. The maker experience simply defaults the source and target branches for the user so they don't have to choose.
 
What is the "proper" process to set up a new development environment from within the ALM Accelerator app? If I created a brand new environment in Power Apps and used the native functionality to import a solution, it would prompt me to set up connections, set environment variables, etc., but importing from source control in the app doesn't do that. You can select a deployment profile when importing, but understandably don't seem to work when importing into an environment other than the one they were created for (since the selected connections don't actually exist).
 - This one is tricky and don't really have a good answer for this today unless the unmanaged solution already exists in another environment. In which case you can create dev stages in your solution profile to create deployment configurations. It's on our backlog to document this further.

Do you recommend "feature" environments to correlate with [feature branches](https://learn.microsoft.com/en-us/azure/devops/repos/git/git-branching-guidance?view=azure-devops) or is it the intention that a maker would have a single environment and delete / re-import solutions as needed to essentially "check out" a feature branch if they are working on multiple simultaneously?
 - Don't have a recommendation, either option is valid. However, the Delete / Import functionality in the ALM Accelerator is meant to facilitate the latter scenario as you mentioned.

On step 5, you mean once the pull request is merged into the main branch, not when the pull request is created, correct?
 - Correct

Are these branch policies actually configured or are there additional steps that need to be completed to make the automatic deployment happen? I don't see any reference to this in the [ALM Accelerator pipeline documentation](https://learn.microsoft.com/en-us/power-platform/guidance/coe/almaccelerator-pipelines).
 - All of the triggers are specified in the pipeline YAML (https://github.com/microsoft/coe-alm-accelerator-templates/blob/main/Pipelines/build-deploy-validation-SampleSolution.yml, https://github.com/microsoft/coe-alm-accelerator-templates/blob/main/Pipelines/build-deploy-test-SampleSolution.yml, https://github.com/microsoft/coe-alm-accelerator-templates/blob/main/Pipelines/build-deploy-prod-SampleSolution.yml)
 - The validation build is triggered by a branch policy that is created when the solution branch is created during the first export of the solution.

**Comment by sefa-phaag** [comment]
Mike, thanks so much for all your answers. One more, though (of course): is the deployment to production in the default configuration immediate? I'm assuming so as I didn't see anything about scheduling it, but it would be nice to be able to approve a pull request during working hours but have the actual deployment wait until the middle of the night sometime.

**Comment by CoEStarterKitBot** [comment]
@sefa-phaag This has been fixed in the latest release. Please install [the latest version of the toolkit](https://aka.ms/CoEStarterKitDownload) following the instructions for [installing updates](https://docs.microsoft.com/power-platform/guidance/coe/setup#installing-updates). Note that if you do not remove the unmanaged layers as described there you will not receive updates from us.
